### ISSUE TITLE
Allow examples to be built at the same time as the main library

### ISSUE THREAD

**Issue by habemus-papadum** [dealii/dealii#3983]
Hello -- currently examples are built by cd'ing into the relevant example directory, and then running cmake.  Implicit in this flow is the prereq that deal.ii has already been built/installed -- i.e. there is a two step process, build deall.ii first and then each example.  This feels suboptimal for a few reasons:

  + examples are one of the core ways of documenting deal.ii and offer are a very quick way to understand holistically how the library works.  This virtue is even further amplified by using smart IDE's that let one use the examples as launching pads to navigate to parts of the codebase while stepping through code via an integrated debugger. 

+ However, the two step build flow currently makes cmake aware ide's such as qt creator and clion work less than optimally.  

For instance if I hack step-1's CMakeList.txt as follows:

````
SET(TARGET "step-1")
SET(TARGET_SRC
  ${TARGET}.cc
  )
CMAKE_MINIMUM_REQUIRED(VERSION 2.8.8)
PROJECT(${TARGET})
add_executable(${TARGET} ${TARGET_SRC})
FOREACH(build ${DEAL_II_BUILD_TYPES})
  DEAL_II_INSOURCE_SETUP_TARGET(${TARGET} ${build})
endforeach()
````   
(coupled with the appropriate `add_subdirectory(step-1)` in the parent CMakeLists.txt) 
then exploration of the code base using clion and debugger becomes substantially better, much faster to setup (which this change, one could simply git clone, point clion to the main CMakeLists.txt, and then with one mouse click build and start stepping thru step-1 in a debugger while having fully intelligent code navigation -- all in about 5 minutes ), etc.  Also this setup makes it easy for me iterate hacking of not just the examples but also the core library code, which is an excellent way to learn how something works.    

If possible, consider adding cmake flags that allow one to choose between the current two step build process or having all examples built alongside the library  using  `in_source` macro.  

Cheers! nehal

**Comment by drwells** [comment]
This sounds like a reasonable thing to do (and I believe that, if we do this, we can get rid of the dummy target we have in place anyway). Can anyone see a downside to this?

People tend to copy and paste the tutorial `CMakeLists.txt` files for their own projects, so we should make sure that, if we do this, we document the current file somewhere explicitly.

**Comment by tjhei** [comment]
An alternative would be to install a CMakeLists.txt in the parent folder (``examples/``) that would configure the projects in the subfolders if requested.

**Comment by bangerth** [comment]
I like the idea of installing the already compiled tutorial programs. I'm going to note that I think this is a bit complicated because the installed executable has to link against the *installed* library, not the one in the build directory. The simplest way is to build the examples during the install phase, but that has the downside that the install phase may frequently be done as root, and we don't want to do too much work as root -- ideally just copying files.

I'm sure @tamiko will come up with a snazzy cmake command that builds the examples and during the install phase just re-links them against the correct library.

**Comment by tjhei** [comment]
> Do we want to teach users of the library to use our compiled binaries?
> 
> The idea behind our example steps is that they are essentially a starting point for your own project - I.e. to modify code. And the only sensible way of doing that I see is to use an independent project. (Think about installing deal.II via your linux distribution package manager.)

I don't think compiled examples should be part of binary packages and I agree that there is value to teach users to go into an example directory and do "cmake .; make run" because they will need to do the same for their projects.

**Comment by habemus-papadum** [comment]
@bangerth 
re: retaining the ability to install examples, making sure executable links to the right library:
This might already be well known to this group, but cmake has excellent support for rpath for exactly this reason (i.e. to avoid having a two step build).  a typical workflow I see is that during a build,  the executable is linked to the library in the build tree with an appropriate (temporary) rpath to the build-tree library, but then on installation of the executable, its rpath is rewritten to reflect the install location of the library (it uses install_name_tool on os x, and "serious monkey business" on linux, but I'm not sure how it works on windows)  The whole process is specified in a "declarative way" and then cmake does an excellent job in making sure it actually works as it should despite how dangerous it seems... 

@tamiko Finding auxiliary input files is indeed a perennial problem -- the approach I have used in the past is to add symlinks in the build dir pointing back to the source tree, but on install to actually copy files into the right location  -- this requires a little cmake effort but, in the end it ends up being two 10 line macros (one for creating symlinks during the build, and one for copying the files during install).  This is a place where I typically resort to placing aux files into a "resources" subdir and use globs to simplify maintenance. 

@tjhei 's idea of implementing this functionality in the `example/`'s CMakeLists.txt seems like the most unobtrusive  way to implement the the desired behavior and is indeed what I ended up doing for the rest of the examples.   
------
In any case, I appreciate the the quick feedback.  I came across one IDE issue (with clion) that I need to submit a bug report about, but once that is resolved, I think I might do a quick screen capture to show how cool it is to be able to hack/navigate/debug the full source of both the examples and the main library  (a header only view of code -- even in the case of such nicely documented code as deal.ii -- just isn't the same... )(ultimately this is the end the goal of this suggestion)

**Comment by tamiko** [comment]
@bangerth No worries, CMake solves the rpath problem quite nicely.


> this requires a little cmake effort but, in the end it ends up being two 10 line macros (one for creating symlinks during the build, and one for copying the files during install). This is a place where I typically resort to placing aux files into a "resources" subdir and use globs to simplify maintenance.

As for your suggestions with symlinks - no that doesn't work. It breaks in-source builds.

I thought about this idea for a bit longer.

Developing a user project with deal.II source in place is by far not the only use-case we support.
And I am not entirely sure how much we gain by directly integrating the examples into the source base.

For example, we have quite a fancy (and absolutely over-engineered) "autopilot" mode for example steps. That won't be available for in-source examples. Further, in-source obviously doesn't help for deal.II being distributed via a package manager (concerns debian/ubuntu/gentoo, homebrew, spack, candi and others).


That being said, I think some sort of modules feature might actually prove quite useful.
What I mean by that would be simply 5 lines in the top-level CMakeLists.txt file that glob over all directories starting with `module-` and include them.

That way you can add arbitrary libraries/programs to the library by simply putting your code into a subdirectory `module-`.

That would allow you to do exactly what you want to do, i.e. have a uniform code view of the library and all your deal.II based projects.

**Comment by tjhei** [comment]
> What I mean by that would be simply 5 lines in the top-level CMakeLists.txt file that glob over all directories starting with module- and include them.

Arguments against that:
- complicated and we need to maintain that
- it is something "custom" and not standard cmake behavior, so we need to document and explain it

**Comment by tjhei** [comment]
> I think I might do a quick screen capture to show how cool it is to be able to hack/navigate/debug the full source of both the examples and the main library

Nice, that would be helpful! We can find a place to link to videos somewhere. I always thought of doing something similar for my workflow (you can do the same using qtcreator), but I never find time for things these days...

**Comment by tamiko** [comment]
@tjhei 
Quite a number of cmake projects do. Trilinos CMake is written with modular packages in mind. The LLVM project does exactly the same thing. And it is not a terribly complicated logic.

Have a look at the pr (1 minute).

**Comment by tamiko** [comment]
@tjhei Yes, I underestimated the code complexity by 60% (I wanted to have a check for the subfolder containing CMakeLists.txt and a status message.)

**Comment by habemus-papadum** [comment]
I should clarify -- I am not interested developing user projects inside of your tree -- that definitely has the feel of an anti-pattern that should be discouraged for anything besides adhoc explorations.  

My point is more along the lines that you could make it easier for your users to get even more out of your examples if they were integrated into the make cmake build -- I think a screen capture will clarify the point quickly

**Comment by habemus-papadum** [comment]
I've made a fork which has the quick patch I use when exploring your examples:
https://github.com/habemus-papadum/dealii/blob/51ddd00f4ce8314078cbe0d2b31cf97756c87709/examples/CMakeLists.txt#L66

Again, to emphasize, my point is that with this patch I am able to learn much more about how dealii works than with the current build flow (because now my IDE let's me hack and navigate and debug anything I want -- I don;t want to just learn how to use dealii, I want to know how it actually works, if it is well written (of course it is!), etc)

**Comment by bangerth** [comment]
@habemus-papadum -- out of curiosity, what exactly is it that this would enable? The ability to ask the IDE for a list of all of the places where a library function is used?

@tamiko -- what if we built each example program (if possible based on dependencies) and then installed the executable? That would indeed simplify the process of running and checking these programs from a user perspective. If they want to modify anything, they will still need to do `cmake . && make`.

In other words, the executable is the only other thing that gets installed, no intermediate build files.

**Comment by habemus-papadum** [comment]
@bangerth 
yes, the IDE's use-def graph becomes more accurate.  

But also, I often make edits to both examples and library code simultaneously (basically git branches are like digital photographs -- there is no cost to experimenting with arbitrarily grotesque hacks -- though quite often I am just adding timing statements ) or set breakpoints in arbitrary locations within the library code so I can inspect the values contained in various internal data structures -- I find watching values change and making sure I understand why they changed the best way to learn.

**Comment by tjhei** [comment]
> edits to both examples and library code simultaneously
> or set breakpoints in arbitrary locations within the library code

Maybe I am missing something but I can open the deal.II project and for example the step-1 project at the same time in qtcreator and all of that just works without any changes.

**Comment by habemus-papadum** [comment]
@tjhei 
I think you are have probably identified the main source of disconnect.    Although the behavior you have described in qtcreator when having two projects open is pretty good -- from my perspective this is still less than ideal.  In this scenario, qt creator creates two distinct use-def graphs (one for each project) that happens to be able to look into the headers of each project creating a pretty good illusion of intelligence.  To further complicate things, the way breakpoints work are very general (the ide models them at the file + line level and are fairly project agnostic ) so indeed having two projects open or even a random source file  and adding breaking points will "just work".  

with that said, I still maintain that the qt creator flow you describe is less than optimal --  I don't believe there is any scenario where you can jump in one step from a .cc in an example to the implementation in a library .cc file (qtcreator keeps getting smarter, so please correct me if I am wrong about its current powers).  in my setup, there have been cases where I select an identifier (e.g. a concrete fe implementation) in step-3, press a few keys that bring up transient heads up displays listing various info (list the based classes of this fe, list all other fe implementations,then list all usages of some other fe), and then when I'm done, I might now be in step-40 -- all of this happening in 2-3 seconds.

**Comment by tjhei** [comment]
> I don't believe there is any scenario where you can jump in one step from a .cc in an example to the implementation in a library .cc file (qtcreator keeps getting smarter, so please correct me if I am wrong 

This works just fine for me: You press F2 and it takes you to the symbol in the header, you press F2 again and you are at the implementation.

**Comment by bangerth** [comment]
Same in Eclipse. As long as both the tutorial and the library are set up as projects in eclipse, it's happy to keep jumping from example to header to source file. If I only set up the tutorial program, then it will still find the header file but of course will no longer know where the source file is.

**Comment by habemus-papadum** [comment]
That's quite good to know;  in this respect both ide's are more powerful than the setup I am using.   Currently the tools I use are clion or vim and various code analysis tools based off of clang's libtooling, all of which are well designed to consume and comprehend a single cmake root, but don't  have much (really, any) ability to bridge information across "projects" -- so in my case I ended up modifying your examples cmake to fit my tools;  I have to assume most of your users will either be using eclipse or qtcreator, and given that they work smoothly with multiple projects, my suggestion may not be so useful.

**Comment by tjhei** [comment]
> so in my case I ended up modifying your examples cmake to fit my tools

I am not totally against making a change similar to what you propose, if it improves the workflow. One problem is of course that such a change won't help with working on a code, which is not one of the examples. I would think that the majority of users work by copying one of the examples to some other location and very few just modify examples. This is at least how I teach it.

**Comment by bangerth** [comment]
That's how I teach it as well, but I see value in being able to just execute tutorials after installation if you just want to play with them.
